### PR TITLE
feat: pull in and convert vscode native html component definitions

### LINKED ISSUE #3772: Use VS Code Web Custom Data package for HTML definitions
# Abstract

Currently the Component Definitions use the same structure as the [VS Code custom data format](https://github.com/microsoft/vscode-custom-data). In order to expand on this so include all standard HTML elements for the sites Component Explorer and Creator the definitions must be available

## Requirements

To keep up-to-date with the latest web standards a dependency should be taken on [VS Code Web Custom Data](https://www.npmjs.com/package/vscode-web-custom-data). This dependency should include definition files for all HTML elements.

- Include the above npm package in the `sites/site-utilities`
- Remove the definitions from the `src/definitions/native` folder and replace them with exports from [VS Code Web Custom Data](https://www.npmjs.com/package/vscode-web-custom-data)
- Test the Creator and Component Explorer sites to ensure this is working as intended.

Note: In the event that this creates so many attributes that the `Form` becomes unweildy to use, instead create custom definitions for native elements for now, open a new task for that and link this one. We may need to complete this task after #2575

**Further edits**
## Additional notes

The `"slots"` are an added part to the schema and were taken as inspiration from https://github.com/webcomponents/custom-elements-json/issues/3, there is a file located in the `@microsoft/fast-tooling`, `data-utilities/html-elements.ts` which includes a string array of the elements that do not contain content. As part of this task, use this array to selectively add the default slots to all other HTML elements.

### PR DESCRIPTION
**Pull request by marjonlynch**
# Description

Add a dependency to vscode's exported html web data definitions and import them was webcomponent definitions for use in fast-component-explorer and fast-creator. As new definitions are added or attributes on existing ones are changed, we just need to get the latest version of vscode's exported package

closes #3772 

## Issue type checklist

- [ ] **Chore**: A change that does not impact distributed packages.
- [ ] **Bug fix**: A change that fixes an issue, link to the issue above.
- [x] **New feature**: A change that adds functionality.

**Is this a breaking change?**
- [ ] This change causes current functionality to break.

**Adding or modifying component(s) in `@microsoft/fast-components` checklist**

- [ ] I have added a new component
- [ ] I have modified an existing component
- [ ] I have updated the [definition file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#definition)
- [ ] I have updated the [configuration file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#configuration)

## Process & policy checklist

- [ ] I have added tests for my changes.
- [x] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/community/code-of-conduct/#our-standards) for this project.

### CODE REVIEW

File: packages/tooling/fast-tooling/src/data-utilities/mapping.vscode-html-languageservice.ts [hunk 1]
```diff
@@ -28,41 +28,43 @@ function mapAttributesAndSlotsToData(
...
-            if (schemaDictionary[schemaId].properties[attributeKey]) {
-                if (
-                    schemaDictionary[schemaId].properties[attributeKey].type ===
-                    DataType.boolean
-                ) {
-                    // When the attribute is a boolean, it does not matter
-                    // what it's value, it will always be true if present
-                    return [attributeKey, true];
-                }
+    if (schemaId !== null && schemaId !== undefined) {
```
Comment (marjonlynch) [inline packages/tooling/fast-tooling/src/data-utilities/mapping.vscode-html-languageservice.ts:31]: Probably didn't need to change this file, will undo this change, check it and push up

File: sites/site-utilities/src/definitions/native/html-native.definition.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
+import { DataType } from "@microsoft/fast-tooling";
+import { WebComponentAttribute } from "@microsoft/fast-tooling/dist/data-utilities/web-component";
+import { voidElements } from "@microsoft/fast-tooling/dist/data-utilities/html-element";
+
+/**
+ * These definitions are dependent on a devDependency on "vscode-web-custom-data"
+ * speficially the browsers.html-data.json which defines native html elements and their attributes
```
Comment (janechu) [inline sites/site-utilities/src/definitions/native/html-native.definition.ts:7]:
```suggestion
 * specifically the browsers.html-data.json which defines native html elements and their attributes
```

File: sites/site-utilities/src/definitions/native/html-native.definition.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+ * as used by vscode
+ */
+import * as vcodeHTMLData from "vscode-web-custom-data/data/browsers.html-data.json";
+
+const valueSets: any[] = vcodeHTMLData.valueSets;
+
+export const htmlNativeDefinitions: any = {
+    version: 1, // WebComponentDefinition has this typed as 1, vcodeHTMLData.version could be anything
```
Comment (janechu) [inline sites/site-utilities/src/definitions/native/html-native.definition.ts:15]:
```suggestion
    version: vsCodeHTMLData.version, // WebComponentDefinition has this typed as 1, vcodeHTMLData.version could be anything so this will need to check for different versions to determine how to approach the conversion
```

File: sites/site-utilities/src/definitions/native/html-native.definition.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
+import { DataType } from "@microsoft/fast-tooling";
+import { WebComponentAttribute } from "@microsoft/fast-tooling/dist/data-utilities/web-component";
+import { voidElements } from "@microsoft/fast-tooling/dist/data-utilities/html-element";
+
+/**
+ * These definitions are dependent on a devDependency on "vscode-web-custom-data"
+ * speficially the browsers.html-data.json which defines native html elements and their attributes
+ * as used by vscode
+ */
+import * as vcodeHTMLData from "vscode-web-custom-data/data/browsers.html-data.json";
```
Comment (janechu) [inline sites/site-utilities/src/definitions/native/html-native.definition.ts:10]:
```suggestion
import * as vscodeHTMLData from "vscode-web-custom-data/data/browsers.html-data.json";
```

File: sites/site-utilities/src/definitions/native/html-native.definition.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+const valueSets: any[] = vcodeHTMLData.valueSets;
+
+export const htmlNativeDefinitions: any = {
```
Comment (janechu) [inline sites/site-utilities/src/definitions/native/html-native.definition.ts:14]: Can you create an interface for vscode version 1.1?

File: sites/site-utilities/src/definitions/native/html-native.definition.ts [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    version: 1, // WebComponentDefinition has this typed as 1, vcodeHTMLData.version could be anything
+    tags: vcodeHTMLData.tags,
+};
+
+function getDataTypeForAttribute(attribute: any): DataType {
```
Comment (janechu) [inline sites/site-utilities/src/definitions/native/html-native.definition.ts:19]: Similarly here, can you create an interface for the attribute? This will help us in the future, you can just create a `html-native.vs-code-v1.1-types.ts` file to store these in.

File: sites/site-utilities/src/definitions/native/html-native.definition.ts [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    /**
+     * For some reason valueSets references "v" in many boolean like attributes but does not define "v",
+     * It does define "b" and both are utilized as booleans for all attributes
+     */
+    switch (attribute.valueSet) {
+        case "b":
+        case "v":
```
Comment (janechu) [inline sites/site-utilities/src/definitions/native/html-native.definition.ts:26]: If you define this as an interface (as I suggested above), you can then set up something like:

```ts
export interface VSCodeAttribute {
    valueSet: "b" | "v"; // add whichever strings are available for valueSet
}
```

File: sites/site-utilities/src/definitions/native/html-native.definition.ts [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+            return DataType.boolean;
+        default:
+            return DataType.string;
+    }
+}
+
+function findValueSetValues(valueSetName: string): any | undefined {
```
Comment (janechu) [inline sites/site-utilities/src/definitions/native/html-native.definition.ts:33]: We should define all of these `any` types.

File: sites/site-utilities/src/definitions/native/html-native.definition.ts (hunk 7 above)
Comment (marjonlynch) [reply sites/site-utilities/src/definitions/native/html-native.definition.ts:31]: Do we want to define the 25 or so strings that are in the definition file for this when we only care about 2?

Comment (janechu) [reply sites/site-utilities/src/definitions/native/html-native.definition.ts:38]: I think leaving it at 2 with a note on where these can be found in the git repository and that this is only a subset would be fine.

File: sites/site-utilities/src/definitions/native/html-native.definition.ts (hunk 5 above)
Comment (marjonlynch) [reply sites/site-utilities/src/definitions/native/html-native.definition.ts]: setting it to 1.1 seems to be a problem, 1 is fine if it matched but 1.1 doesn't seem to work

Comment (janechu) [reply sites/site-utilities/src/definitions/native/html-native.definition.ts]: That's because you didn't create your own definition, you used the toolings `WebComponentDefinition` which is strongly typed at `1`. What I'm suggesting is creating your own vscode definition for 1.1 instead of using mine, since they have differences and this will help us track that.

Comment (marjonlynch) [reply sites/site-utilities/src/definitions/native/html-native.definition.ts]: I think we're talking about this which is my own definition, AFAIK
`
export interface VSCodeNativeHTMLDefinition {
    version: number;
    tags: VSCodeNativeHTMLTag[];
}
`

Comment (marjonlynch) [reply sites/site-utilities/src/definitions/native/html-native.definition.ts]: Screen Shot 2020-11-04 at 8 56 54 AM

File: sites/site-utilities/src/definitions/native/html-native.definition.ts (hunk 7 above)
Comment (marjonlynch) [reply sites/site-utilities/src/definitions/native/html-native.definition.ts:39]: done and done

File: sites/site-utilities/src/definitions/native/html-native.definition.ts (hunk 5 above)
Comment (marjonlynch) [reply sites/site-utilities/src/definitions/native/html-native.definition.ts]: I don't know why that doesn't work, possibly version of typescript we are using but I'm just casting it in the import part and I have the version set at 1.1 and it's working fine.

**Review by janechu — APPROVED** [review]

**Review by EisenbergEffect — APPROVED** [review]
